### PR TITLE
feat(658): Deploy llmd and evaluate

### PR DESCRIPTION
**Pull request by Dericko681**
## 1. Summary

This PR changes:

- Deploy llm with vllm and a test model

It solves:

- #658 

---

## 2. Intent

The intent of this PR is:

> Deploy the llmd and play around with

---

### In Scope

- [Included change 1]
- [Included change 2]

### Out of Scope

- [Excluded change 1]
- [Excluded change 2]

---

## 4. Verification

I verified this change by:

- [ ] Running automated tests
- [x] Running manual tests
- [ ] Checking logs
- [ ] Checking metrics
- [ ] Testing error cases
- [ ] Testing permissions/security behavior
- [ ] Testing rollback or failure behavior, if relevant

Commands run:

```sh
curl -s --max-time 30 -X POST http://localhost:18081/v1/completions \
    -H 'Content-Type: application/json' \
    -d '{"model":"Qwen/Qwen2.5-0.5B-Instruct","prompt":"What is the capital of cameroon?","max_tokens":10}' | jq
 ```
```sh
open plans/artillery/reports/benchmark-results.html
```

Results:
image

```text
curl -s --max-time 30 -X POST http://localhost:18081/v1/completions \
    -H 'Content-Type: application/json' \
    -d '{"model":"Qwen/Qwen2.5-0.5B-Instruct","prompt":"What is the capital of cameroon?","max_tokens":10}' | jq
{
  "id": "cmpl-e644cea3-f2ed-48e4-b034-264e7305c095",
  "object": "text_completion",
  "created": 1784134076,
  "model": "Qwen/Qwen2.5-0.5B-Instruct",
  "choices": [
    {
      "index": 0,
      "text": " Cameroon's capital city is Yaounde. Ya",
      "logprobs": null,
      "finish_reason": "length",
      "stop_reason": null,
      "token_ids": null,
      "prompt_logprobs": null,
      "prompt_token_ids": null,
      "routed_experts": null
    }
  ],
  "service_tier": null,
  "system_fingerprint": "vllm-0.25.1-2b261bc6",
  "usage": {
    "prompt_tokens": 8,
    "total_tokens": 18,
    "completion_tokens": 10,
    "prompt_tokens_details": null
  },
  "kv_transfer_params": null,
  "metrics": null
}
```
---

## 5. Screenshots / Evidence

image

* Screenshot: [link]
* Logs: [link]
* Metrics: [link]
* Recording: [link]

---

## 6. Risk Assessment

Risk level:

* [x] Low
* [ ] Medium
* [ ] High

Potential risks:

* [Risk 1]
* [Risk 2]

Mitigation:

* [Mitigation 1]
* [Mitigation 2]

---

## 7. AI Usage Declaration

AI was used for:

* [x] Understanding existing code
* [x] Generating code
* [ ] Refactoring
* [ ] Generating tests
* [ ] Drafting documentation
* [ ] Reviewing the diff
* [ ] Not used

Human verification:

* [ ] I understand every meaningful change in this PR
* [ ] I checked generated code manually
* [ ] I checked generated tests manually
* [ ] I removed unsupported AI assumptions
* [ ] I accept responsibility for this PR

---

## 8. Reviewer Focus

Please focus your review on:

* [x] Correctness
* [ ] Architecture
* [ ] Security
* [ ] Performance
* [ ] Tests
* [ ] Maintainability
* [ ] Product intent
* [ ] Edge cases

### CODE REVIEW

**Review by stephane-segning — COMMENTED** [review]
I'm not sure I understand this very well. So this means that the perf is worse?

Image

Bon I guess we need more GPUs and more instances to make this investigation more legit, right?

File: charts/llm-d/values.yaml [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    main:
+      type: statefulset
+      replicas: 1                # one GPU -> one replica (multi-GPU: bump this)
+      annotations:
+        argocd.argoproj.io/sync-wave: '1'
+      containers:
+        model:
+          image:
+            repository: vllm/vllm-openai
+            tag: latest           # local k3d PoC — pin for production
```
Comment (stephane-segning) [inline charts/llm-d/values.yaml]: **`tag: latest` + `pullPolicy: IfNotPresent` is a non-reproducible pairing.** `latest` is a moving target (the benchmark in this PR isn't reproducible against a named vLLM version), and `IfNotPresent` means each node keeps whatever `latest` it first cached and never re-pulls — so different nodes/re-deploys can silently run different vLLM builds. The comment says "pin for production," but even for a PoC an explicit tag (e.g. `v0.11.0`) makes the evaluation reproducible. Pin it, or at minimum use `pullPolicy: Always` with `latest`.

File: charts/llm-d/values.yaml [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+          args:
+          - 'MODEL_DIR="/models/Qwen2.5-0.5B-Instruct"
+
+            mkdir -p "$MODEL_DIR"
+
+            echo "Seeding Qwen/Qwen2.5-0.5B-Instruct -> $MODEL_DIR"
+
+            pip install --no-cache-dir -q --root-user-action=ignore "huggingface_hub[hf_xet]>=1.0"
+
+            hf download Qwen/Qwen2.5-0.5B-Instruct --local-dir "$MODEL_DIR"
```
Comment (stephane-segning) [inline charts/llm-d/values.yaml:327]: **Seed download is hardcoded and decoupled from `model.*`.** `model.hfRepo` (`Qwen/Qwen2.5-0.5B-Instruct`) and `model.storagePath` exist above, but this seed Job hardcodes both the repo and `/models/Qwen2.5-0.5B-Instruct`. Changing `model.hfRepo`/`model.storagePath` silently does NOT change what gets downloaded — the vLLM container would then mount an empty/wrong subPath and crashloop. The inline comment acknowledges the bjw scope limitation, but this is a real drift trap: consider passing the repo/path via a subchart-scoped value (bjw supports per-controller env/values) or asserting equality, rather than relying on a "keep in sync" comment.

File: charts/llm-d/values.yaml [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+global:
+  # Used by the chart OWN templates labels (common.labels.standard).
+  labels:
+    app: llm-d
+    team: adorsys-gis
+
+model:
+  # Logical model name — the pod label EPP selects on (--endpoint-selector).
+  # Also the StatefulSet/Service name suffix (via modelServing.fullnameOverride).
+  name: llm-d-vllm
```
Comment (stephane-segning) [inline charts/llm-d/values.yaml:40]: **`model.name` is decorative — it doesn't actually drive the selector or the name.** The comment says it's "the pod label EPP selects on" and "the StatefulSet/Service name suffix," but the EPP selector (`llmdRouter.router.modelServers.matchLabels.app`), the pod label (`modelServing.defaultPodOptions.labels.app`), and `modelServing.global.fullnameOverride` are all independently **hardcoded** to `llm-d-vllm`. Changing `model.name` changes none of them → the label selector silently stops matching the pods. Either template those three from `model.name` (`{{ .Values.model.name }}`) so it's a single source of truth, or drop the misleading comment.

File: charts/llm-d/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  The chart own templates/ render the optional HF-token ExternalSecret (for
+  gated models). The vLLM pods carry app: <model.name> so EPP
+  --endpoint-selector discovers them.
+
+  llm-d: https://llm-d.ai  |  GAIE: gateway-api-inference-extension
+type: application
+version: 0.1.0
+appVersion: "v0.9.0"
+kubeVersion: ">=1.28.0-0"
+keywords:
```
Comment (stephane-segning) [inline charts/llm-d/Chart.yaml:27]: **Not registered in `charts/apps/values.yaml`.** This chart has no entry in the ArgoCD app-of-apps `applications:` list, so it won't deploy via GitOps — it's `helm install`-only (matching the k3d PoC framing). That's fine for an "evaluate" ticket, but worth confirming it's intentional and noting in the README that this chart is **not** part of the deployed fleet (so a future reader doesn't assume ArgoCD manages it). If it's meant to run on the cluster, it needs an app entry + the ADR-0017 destination wiring like the other model-serving charts.

**Review by Koufan-De-King — APPROVED** [review]
